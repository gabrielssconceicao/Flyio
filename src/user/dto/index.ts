import { CreateUserDto } from './create-user.dto';
import { FindAllUsersResponseDto } from './find-all-users.dto';
import { QueryParamDto } from './query-param.dto';
import { ReactivateUserDto } from './reactivate-user.dto';
import { UpdateUserDto } from './update-user.dto';

export {
  CreateUserDto,
  FindAllUsersResponseDto,
  QueryParamDto,
  ReactivateUserDto,
  UpdateUserDto,
};
