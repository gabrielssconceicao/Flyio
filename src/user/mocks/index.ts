import { generateUserMock } from './user.mock';
import { generateFindAllUsersResponseDtoMock } from './generate-find-all-users-response.dto.mock';
import { generateCreateUserDtoMock } from './generate-create-user.dto.mock';

export {
  generateUserMock,
  generateFindAllUsersResponseDtoMock,
  generateCreateUserDtoMock,
};
