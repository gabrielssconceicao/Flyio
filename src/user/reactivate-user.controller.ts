import { Controller, HttpCode, HttpStatus, Param, Patch } from '@nestjs/common';
import { ApiBody, ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { ReactivateUserService } from './reactivate-user.service';
import { ReactivateUserDto } from './dto';

@ApiTags('Users')
@Controller('users')
export class ReactivateUserController {
  constructor(private readonly reactivateUserService: ReactivateUserService) {}

  @ApiOperation({ summary: 'Reactivate user by token' })
  @ApiBody({
    type: ReactivateUserDto,
    description: 'Reactivate user by token',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Account reactivated successfully.',
    schema: {
      example: {
        message: 'Account reactivated successfully',
      },
    },
  })
  @ApiResponse({
    status: HttpStatus.UNAUTHORIZED,
    description: 'Invalid or expired token.',
    schema: {
      example: {
        message: 'Invalid or expired token',
      },
    },
  })
  @HttpCode(HttpStatus.CREATED)
  @Patch('reactivate/:token')
  reactivate(@Param() reactivateUserDto: ReactivateUserDto) {
    return this.reactivateUserService.reactivate(reactivateUserDto);
  }
}
