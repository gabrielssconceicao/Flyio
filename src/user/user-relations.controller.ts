import {
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  Patch,
  Query,
  UseGuards,
} from '@nestjs/common';
import {
  ApiBody,
  ApiOperation,
  ApiParam,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { ReactivateUserDto } from './dto';
import { UserRelationsService } from './user-relations.service';
import { PaginationDto } from '../common/dto/pagination.dto';
import { AuthTokenGuard } from '../auth/guard/auth-token.guard';
import { ApiAuthResponses } from '../common/decorators/guard.decorator';
import { FindAllUsersPostsResponseDto } from './dto/find-all-user-posts.dto';

@ApiTags('Users')
@Controller('users')
export class UserRelationsController {
  constructor(private readonly userRelationsService: UserRelationsService) {}

  @ApiOperation({ summary: 'Reactivate user by token' })
  @ApiBody({
    type: ReactivateUserDto,
    description: 'Reactivate user by token',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Account reactivated successfully.',
    schema: {
      example: {
        message: 'Account reactivated successfully',
      },
    },
  })
  @ApiResponse({
    status: HttpStatus.UNAUTHORIZED,
    description: 'Invalid or expired token.',
    schema: {
      example: {
        message: 'Invalid or expired token',
      },
    },
  })
  @HttpCode(HttpStatus.CREATED)
  @Patch('reactivate/:token')
  reactivate(@Param() reactivateUserDto: ReactivateUserDto) {
    return this.userRelationsService.reactivate(reactivateUserDto);
  }

  @ApiOperation({ summary: 'Get all posts of a user by username' })
  @ApiParam({
    name: 'username',
    description: 'Username of the user',
    example: 'jDoe453',
  })
  @ApiResponse({
    status: HttpStatus.OK,
    description: 'List of posts of a user with limited details',
    type: FindAllUsersPostsResponseDto,
  })
  @ApiAuthResponses()
  @HttpCode(HttpStatus.OK)
  @UseGuards(AuthTokenGuard)
  @Get(':username/posts')
  getAllPostsByUsername(
    @Param('username') username: string,
    @Query() paginationDto: PaginationDto,
  ) {
    return this.userRelationsService.getAllPostsByUsername(
      username,
      paginationDto,
    );
  }
}
