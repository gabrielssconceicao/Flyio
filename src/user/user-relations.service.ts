import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { FindAllUsersResponseDto } from './dto';
import { PaginationDto } from '../common/dto/pagination.dto';
import { FindAllUsersPostsResponseDto } from './dto/find-all-user-posts.dto';
import { FindAllLikedPostsResponseDto } from '../post/dto/find-all-liked-post.dto';
import { TokenPayloadDto } from '../auth/dto';

@Injectable()
export class UserRelationsService {
  constructor(private readonly prismaService: PrismaService) {}

  private selectPostFields = {
    id: true,
    content: true,
    createdAt: true,
    images: {
      select: {
        url: true,
        id: true,
      },
    },
    user: {
      select: {
        username: true,
        profileImg: true,
      },
    },
    _count: {
      select: {
        PostLikes: true,
      },
    },
  };

  async getAllPostsByUsername(
    username: string,
    paginationDto: PaginationDto,
    tokenPayload: TokenPayloadDto,
  ): Promise<FindAllUsersPostsResponseDto> {
    const { limit = 50, offset = 0 } = paginationDto;

    await this.userExists(username);

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.post,
      {
        where: { user: { username } },
        select: {
          ...this.selectPostFields,
          PostLikes: {
            where: { userId: tokenPayload.sub },
            select: { userId: true },
          },
        },
        orderBy: { createdAt: 'desc' },
        skip: offset,
        take: limit,
      },
    );

    const mappedPosts = items.map(({ _count, PostLikes, ...post }) => ({
      ...post,
      likes: _count.PostLikes,
      liked: !!PostLikes.length,
    }));

    return { count, items: mappedPosts };
  }

  async getAllLikedPostsByUsername(
    username: string,
    paginationDto: PaginationDto,
  ): Promise<FindAllLikedPostsResponseDto> {
    const { limit = 50, offset = 0 } = paginationDto;
    const user = await this.userExists(username);

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.post,
      {
        where: {
          PostLikes: {
            some: {
              userId: user.id,
            },
          },
        },
        orderBy: {
          createdAt: 'desc',
        },
        take: limit,
        skip: offset,
        select: this.selectPostFields,
      },
    );

    return {
      count,
      items: items.map(({ _count, ...post }) => {
        return { ...post, likes: _count.PostLikes, liked: true };
      }),
    };
  }

  async followUser(username: string, tokenPayload: TokenPayloadDto) {
    const followingUser = await this.userExists(username);

    if (!followingUser) {
      throw new NotFoundException('User not found');
    }

    if (followingUser.id === tokenPayload.sub) {
      throw new BadRequestException('You cannot follow yourself');
    }

    const follower = await this.prismaService.follower.findFirst({
      where: {
        followingId: followingUser.id,
        userId: tokenPayload.sub,
      },
    });

    if (follower) {
      throw new BadRequestException('User already followed');
    }

    await this.prismaService.follower.create({
      data: {
        followingId: followingUser.id,
        userId: tokenPayload.sub,
      },
    });
  }

  async unfollowUser(username: string, tokenPayload: TokenPayloadDto) {
    const followingUser = await this.userExists(username);

    if (followingUser.id === tokenPayload.sub) {
      throw new BadRequestException('You cannot unfollow yourself');
    }

    const follower = await this.prismaService.follower.findFirst({
      where: {
        followingId: followingUser.id,
        userId: tokenPayload.sub,
      },
    });

    if (!follower) {
      throw new BadRequestException('User already unfollowed');
    }

    await this.prismaService.follower.delete({
      where: {
        userId_followingId: {
          followingId: followingUser.id,
          userId: tokenPayload.sub,
        },
      },
    });
  }
  async getAllFollowingsByUser(
    username: string,
  ): Promise<FindAllUsersResponseDto> {
    const user = await this.userExists(username);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.follower,
      {
        where: { userId: user.id },
        select: {
          following: {
            select: {
              username: true,
              name: true,
              profileImg: true,
              id: true,
            },
          },
        },
      },
    );

    return { count, items: items.map((item) => item.following) };
  }
  async getAllFollowersByUser(
    username: string,
  ): Promise<FindAllUsersResponseDto> {
    const user = await this.userExists(username);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    const { count, items } = await this.prismaService.findAll(
      this.prismaService.follower,
      {
        where: { followingId: user.id },
        select: {
          user: {
            select: {
              username: true,
              name: true,
              profileImg: true,
              id: true,
            },
          },
        },
      },
    );

    return { count, items: items.map((item) => item.user) };
  }

  private async userExists(username: string) {
    const user = await this.prismaService.user.findUnique({
      where: { username },
    });

    if (!user) {
      throw new NotFoundException('User not found');
    }

    return user;
  }
}
