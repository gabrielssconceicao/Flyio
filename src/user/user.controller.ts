import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  HttpStatus,
  HttpCode,
} from '@nestjs/common';
import { UserService } from './user.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { ApiOperation, ApiTags } from '@nestjs/swagger';

@ApiTags('User')
@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}

  /*
  @ApiOperation({ summary: 'Criar um novo recado' }) // Descrição da operação
  @ApiResponse({
    status: 201,
    description: 'Recado criado com sucesso.',
    type: ResponseRecadoDto,
  })
  @ApiResponse({
    status: 400,
    description: 'Dados inválidos.',
    type: BadRequestException,
    example: new BadRequestException('Error Message').getResponse(),
  })
  */
  @ApiOperation({ summary: 'Create a new user' })
  @HttpCode(HttpStatus.CREATED)
  @Post('register')
  create(@Body() createUserDto: CreateUserDto) {
    console.log(createUserDto);
    return this.userService.create(createUserDto);
  }

  @Get()
  findAll() {
    return this.userService.findAll();
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.userService.findOne(+id);
  }

  @Patch(':id')
  update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto) {
    return this.userService.update(+id, updateUserDto);
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.userService.remove(+id);
  }
}
