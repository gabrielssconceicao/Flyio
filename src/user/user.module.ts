import { Module } from '@nestjs/common';
import { UserService } from './user.service';
import { UserRelationsService } from './user-relations.service';
import { UserController } from './user.controller';
import { UserRelationsController } from './user-relations.controller';

@Module({
  imports: [],
  controllers: [UserController, UserRelationsController],
  providers: [UserService, UserRelationsService],
})
export class UserModule {}
