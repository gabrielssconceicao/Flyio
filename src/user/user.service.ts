import {
  ConflictException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { HashingServiceProtocol } from '../auth/hashing/hashing.service';
import { PrismaService } from '../prisma/prisma.service';
import { User } from './entities/user.entity';
import { QueryParamDto } from './dto/query-param.dto';
import { FindAllUsersResponseDto } from './dto/find-all-users.dto';
import { CloudinaryService } from '../cloudinary/cloudinary.service';

@Injectable()
export class UserService {
  constructor(
    private readonly hashingService: HashingServiceProtocol,
    private readonly prismaService: PrismaService,
    private readonly cloudinaryService: CloudinaryService,
  ) {}

  private selectUserFields = {
    id: true,
    name: true,
    username: true,
    email: true,
    profileImg: true,
    bio: true,
    active: true,
  };

  async create(
    createUserDto: CreateUserDto,
    profileImg: Express.Multer.File,
  ): Promise<User> {
    const { password, ...rest } = createUserDto;

    const userExists = await this.prismaService.user.findFirst({
      where: { OR: [{ email: rest.email }, { username: rest.username }] },
    });
    if (!!userExists) {
      throw new ConflictException(
        'This email or username is already associated with an existing account',
      );
    }

    let profileImgUrl = null;
    if (!!profileImg) {
      profileImgUrl =
        await this.cloudinaryService.uploadProfilePicture(profileImg);
    }
    const hashedPassword = await this.hashingService.hash(password);
    const createdUser = await this.prismaService.user.create({
      data: { ...rest, password: hashedPassword, profileImg: profileImgUrl },
      select: this.selectUserFields,
    });
    return createdUser;
  }

  async findAll(query: QueryParamDto): Promise<FindAllUsersResponseDto> {
    const { limit = 50, offset = 0, search = '' } = query;
    const where = search
      ? {
          OR: [{ name: { contains: search } }, { email: { contains: search } }],
        }
      : {};
    const users = await this.prismaService.user.findMany({
      where: where,
      select: {
        id: true,
        name: true,
        username: true,
        profileImg: true,
      },
      take: limit,
      skip: offset,
    });
    const count = await this.prismaService.user.count({ where });
    return { count, users };
  }

  async findOne(username: string): Promise<User> {
    const user = await this.prismaService.user.findUnique({
      where: { username },
      select: this.selectUserFields,
    });
    if (!user) {
      throw new NotFoundException('User not found');
    }

    return user;
  }

  async update(
    username: string,
    updateUserDto: UpdateUserDto,
    profileImg: Express.Multer.File,
  ): Promise<User> {
    const user = await this.findOne(username);

    if (updateUserDto?.email && updateUserDto.email !== user.email) {
      const emailIsTaken = await this.prismaService.user.findFirst({
        where: { email: updateUserDto?.email, NOT: { id: user.id } },
      });

      if (emailIsTaken) {
        throw new ConflictException(
          'Email is already associated with an existing account',
        );
      }
    }

    const personDto = {
      name: updateUserDto?.name,
      email: updateUserDto?.email,
      bio: updateUserDto?.bio,
    };

    if (updateUserDto?.password) {
      const passwordHash = await this.hashingService.hash(
        updateUserDto.password,
      );
      personDto['password'] = passwordHash;
    }

    if (!!profileImg) {
      if (user.profileImg) {
        personDto['profileImg'] =
          await this.cloudinaryService.updateProfilePicture(
            profileImg,
            user.profileImg,
          );
      } else {
        personDto['profileImg'] =
          await this.cloudinaryService.uploadProfilePicture(profileImg);
      }
    }

    const updatedUser = await this.prismaService.user.update({
      where: { username },
      data: { ...personDto },
      select: this.selectUserFields,
    });

    return updatedUser;
  }

  async remove(id: string) {
    const user = await this.prismaService.user.findUnique({
      where: { id },
    });
    if (!user || !user.active) {
      throw new NotFoundException('User not found');
    }

    await this.prismaService.user.update({
      where: { id },
      data: { active: false },
    });
    return { message: 'User deleted successfully' };
  }

  async removeProfilePicture(username: string): Promise<User> {
    const user = await this.findOne(username);

    if (user.profileImg) {
      try {
        await this.cloudinaryService.deleteProfilePicture(user.profileImg);
        const userWithoutProfile = await this.prismaService.user.update({
          where: { id: user.id },
          data: {
            profileImg: null,
          },
          select: this.selectUserFields,
        });
        return userWithoutProfile;
      } catch (error) {
        throw error;
      }
    }
    return user;
  }
}
