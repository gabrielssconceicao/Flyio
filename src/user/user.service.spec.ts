import { Test, TestingModule } from '@nestjs/testing';
import { UserService } from './user.service';
import { HashingServiceProtocol } from '../auth/hashing/hashing.service';
import { PrismaService } from '../prisma/prisma.service';
import {
  selectUserFieldsMock,
  userPrismaService,
} from '../mocks/services/prisma.service.mock';
import { hashingServiceMock } from '../mocks/services/hashing.service.mock';
import { ConflictException, NotFoundException } from '@nestjs/common';
import {
  generateCreateUserDtoMock,
  generateFindAllUsersResponseDtoMock,
  generateUserMock,
} from './mocks/user.mock';
import { CloudinaryService } from '../file/cloudinary.service';
import { cloudinaryServiceMock } from '../mocks/services/cloudinary.service.mock';
import {
  generatedProfilePictureMock,
  generateFileMock,
} from '../mocks/data/file.mock';

describe('<UserService />', () => {
  let service: UserService;
  let hashingService: HashingServiceProtocol;
  let prismaService: PrismaService;
  let cloudinaryService: CloudinaryService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        UserService,
        {
          provide: HashingServiceProtocol,
          useValue: hashingServiceMock,
        },
        {
          provide: PrismaService,
          useValue: userPrismaService,
        },
        {
          provide: CloudinaryService,
          useValue: cloudinaryServiceMock,
        },
      ],
    }).compile();

    service = module.get<UserService>(UserService);
    hashingService = module.get<HashingServiceProtocol>(HashingServiceProtocol);
    prismaService = module.get<PrismaService>(PrismaService);
    cloudinaryService = module.get<CloudinaryService>(CloudinaryService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
    expect(hashingService).toBeDefined();
    expect(prismaService).toBeDefined();
    expect(cloudinaryService).toBeDefined();
  });

  describe('<Create />', () => {
    it('should create a user successfully', async () => {
      const passwordHash = 'HASH_PASSWORD';
      const createUserDto = generateCreateUserDtoMock(true);
      const newUser = generateUserMock();
      const imageFile = generateFileMock();

      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue(null);
      jest
        .spyOn(cloudinaryService, 'uploadProfilePicture')
        .mockResolvedValue(generatedProfilePictureMock);
      jest.spyOn(hashingService, 'hash').mockResolvedValue(passwordHash);
      jest
        .spyOn(prismaService.user, 'create')
        .mockResolvedValue(newUser as any);

      const result = await service.create(createUserDto, imageFile);

      expect(prismaService.user.findFirst).toHaveBeenCalled();
      expect(cloudinaryService.uploadProfilePicture).toHaveBeenCalledWith(
        imageFile,
      );
      expect(hashingService.hash).toHaveBeenCalledWith(createUserDto.password);
      expect(prismaService.user.create).toHaveBeenCalledWith({
        data: {
          ...createUserDto,
          password: passwordHash,
          profileImg: generatedProfilePictureMock,
        },
        select: selectUserFieldsMock,
      });

      expect(result).toEqual(newUser);
      expect(result).toMatchSnapshot();
    });

    it('should create a user successfully without profile picture', async () => {
      const passwordHash = 'HASH_PASSWORD';
      const createUserDto = generateCreateUserDtoMock(true);
      const newUser = generateUserMock();
      const imageFile = null;

      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue(null);
      jest.spyOn(hashingService, 'hash').mockResolvedValue(passwordHash);
      jest
        .spyOn(prismaService.user, 'create')
        .mockResolvedValue(newUser as any);

      const result = await service.create(createUserDto, imageFile);

      expect(prismaService.user.findFirst).toHaveBeenCalled();
      expect(cloudinaryService.uploadProfilePicture).not.toHaveBeenCalled();
      expect(hashingService.hash).toHaveBeenCalledWith(createUserDto.password);
      expect(prismaService.user.create).toHaveBeenCalledWith({
        data: {
          ...createUserDto,
          password: passwordHash,
          profileImg: imageFile,
        },
        select: selectUserFieldsMock,
      });

      expect(result).toEqual(newUser);
      expect(result).toMatchSnapshot();
    });

    it('should throw an ConflictException if user already exists', async () => {
      jest.spyOn(prismaService.user, 'findFirst').mockResolvedValue({} as any);

      await expect(service.create({} as any, null)).rejects.toThrow(
        ConflictException,
      );
    });
  });

  describe('<FindOne />', () => {
    it('should return a user successfully', async () => {
      const username = generateCreateUserDtoMock().username;
      const user = generateUserMock();

      jest
        .spyOn(prismaService.user, 'findUnique')
        .mockResolvedValue(user as any);

      const result = await service.findOne(username);

      expect(prismaService.user.findUnique).toHaveBeenCalledWith({
        where: { username },
        select: selectUserFieldsMock,
      });

      expect(result).toEqual(result);
      expect(result).toMatchSnapshot();
    });

    it('should throw an NotFoundException', async () => {
      jest.spyOn(prismaService.user, 'findUnique').mockResolvedValue(null);

      await expect(service.findOne('jD')).rejects.toThrow(NotFoundException);
    });
  });

  describe('<Update />', () => {
    it('should update a user successfully with profile picture', async () => {
      const updateUserDto = {
        name: 'new name',
        password: 'new_password',
        bio: 'new bio',
      };
      const username = generateCreateUserDtoMock().username;

      const user = generateUserMock();
      const imageFile = generateFileMock();
      const passwordHash = 'HASH_PASSWORD';
      jest.spyOn(service, 'findOne').mockResolvedValue(user as any);
      jest.spyOn(hashingService, 'hash').mockResolvedValue(passwordHash);
      jest
        .spyOn(cloudinaryService, 'uploadProfilePicture')
        .mockResolvedValue(generatedProfilePictureMock);
      jest.spyOn(prismaService.user, 'update').mockResolvedValue({
        ...updateUserDto,
        ...user,
      } as any);

      const result = await service.update(username, updateUserDto, imageFile);
      expect(service.findOne).toHaveBeenCalledWith(username);
      expect(hashingService.hash).toHaveBeenCalledWith(updateUserDto.password);
      expect(cloudinaryService.uploadProfilePicture).toHaveBeenCalledWith(
        imageFile,
      );
      expect(prismaService.user.update).toHaveBeenCalledWith({
        where: { username },
        data: {
          ...updateUserDto,
          password: passwordHash,
          profileImg: generatedProfilePictureMock,
        },
        select: selectUserFieldsMock,
      });
      expect(result).toEqual({
        ...updateUserDto,
        ...user,
      });

      expect(result).toMatchSnapshot();
    });

    it('should throw an error if user not found', async () => {
      jest.spyOn(service, 'findOne').mockRejectedValue(new NotFoundException());

      await expect(
        service.update('jDoe', {} as any, undefined),
      ).rejects.toThrow(NotFoundException);
    });

    it('should throw an error if email is already in use', async () => {
      const updateUserDto = {
        email: 'newemail@example.com',
      };

      const existingUserMock = {
        id: 1,
        username: 'jDoe2',
        email: 'johndoe@example.com',
      };

      const conflictingUserMock = {
        id: 2,
        username: 'otherUser',
        email: 'newemail@example.com',
      };

      jest.spyOn(service, 'findOne').mockResolvedValue(existingUserMock as any);

      jest
        .spyOn(prismaService.user, 'findFirst')
        .mockResolvedValue(conflictingUserMock as any);

      await expect(
        service.update('jDoe2', updateUserDto, undefined),
      ).rejects.toThrow(ConflictException);
    });
  });

  describe('<Delete />', () => {
    it('should delete a user successfully', async () => {
      jest
        .spyOn(prismaService.user, 'findUnique')
        .mockResolvedValue({ active: true } as any);
      jest
        .spyOn(prismaService.user, 'update')
        .mockResolvedValue({ active: false } as any);

      const id = generateUserMock().id;

      const result = await service.remove(id);
      expect(prismaService.user.findUnique).toHaveBeenCalledWith({
        where: { id: id },
      });
      expect(prismaService.user.update).toHaveBeenCalledWith({
        where: { id: id },
        data: { active: false },
      });
      expect(result).toEqual({ message: 'User deleted successfully' });
      expect(result).toMatchSnapshot();
    });

    it('should throw a NotFoundException if user not found', async () => {
      jest
        .spyOn(prismaService.user, 'findUnique')
        .mockResolvedValue({ active: false } as any);
      await expect(service.remove('jD')).rejects.toThrow(NotFoundException);
    });
  });

  describe('<FindAll />', () => {
    it('should search for a list of users ', async () => {
      const query = {
        search: 'jD',
        limit: 1,
        offset: 0,
      };
      const findAllResponse = generateFindAllUsersResponseDtoMock();
      jest
        .spyOn(prismaService.user, 'findMany')
        .mockResolvedValue(findAllResponse.users as any);
      jest
        .spyOn(prismaService.user, 'count')
        .mockResolvedValue(findAllResponse.count);

      const result = await service.findAll(query);

      expect(prismaService.user.findMany).toHaveBeenCalled();
      expect(prismaService.user.count).toHaveBeenCalled();
      expect(result.count).toBe(1);
      expect(result.users.length).toBe(1);
      expect(result).toEqual(findAllResponse);
      expect(result).toMatchSnapshot();
    });
  });
});
