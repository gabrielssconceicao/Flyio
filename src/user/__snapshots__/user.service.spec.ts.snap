// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserService /> <Create /> should create a user successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "email": "jdoe@me.com",
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;

exports[`<UserService /> <Delete /> should delete a user successfully 1`] = `
{
  "message": "User deleted successfully",
}
`;

exports[`<UserService /> <FindOne /> should return a user successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "email": "jdoe@me.com",
  "id": "42-d-f-df4",
  "name": "John Doe",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;

exports[`<UserService /> <Update /> should update a user successfully 1`] = `
{
  "active": true,
  "bio": "This is my bio",
  "email": "jdoe@me.com",
  "id": "42-d-f-df4",
  "name": "John Doe",
  "password": "new_password",
  "profileImg": "https://example.com/profile.jpg",
  "username": "jDoe453",
}
`;
